lib/logstash/inputs/gelf.rb

```diff
@@ -68,36 +76,135 @@ def initialize(params)
   public
   def register
     require 'gelfd'
+    @tcp = nil
+    @udp = nil
+    if @port_tcp == nil
+      @port_tcp = @port
+    end
+    if @port_udp == nil
+      @port_udp = @port
+    end  
   end # def register
 
   public
   def run(output_queue)
     begin
-      # udp server
-      udp_listener(output_queue)
+      if @use_tcp
+        tcp_thr = Thread.new(output_queue) do |output_queue|
+          tcp_listener(output_queue)
+        end
+      end
+      if @use_udp
+        udp_thr = Thread.new(output_queue) do |output_queue|
+          udp_listener(output_queue)
+        end
+      end
     rescue => e
       unless stop?
         @logger.warn("gelf listener died", :exception => e, :backtrace => e.backtrace)
         Stud.stoppable_sleep(RECONNECT_BACKOFF_SLEEP) { stop? }
         retry unless stop?
       end
     end # begin
+    if @use_tcp
+      tcp_thr.join
+    end
+    if @use_udp
+      udp_thr.join
+    end
   end # def run
 
   public
   def stop
-    @udp.close
+    if @use_tcp && @use_udp
+      return @udp.close && tcp.close
+    elseif @use_tcp
+      return @tcp.close
+    elseif @use_udp
+      return @udp.close
+    end
+    return true
   rescue IOError # the plugin is currently shutting down, so its safe to ignore theses errors
   end
 
+  private
+  def tcp_listener(output_queue)
+
+    @logger.info("Starting gelf listener (tcp) ...", :address => "#{@host}:#{@port_tcp}")
+
+    if @tcp.nil?
```

Comment:
Because we use tcp = null above i doint see where it changes the value of this when use_tcp is enabled, same for udp.

Reply:
@tcp is the tcp socket. In register it is set nil. In tcp_listener it is then initialized to @tcp = TCPServer.new(@host, @port_tcp).
I didn't write this code myself and I'm not sure whether this nil initialization and check are necessary but the intention of the author must have been to make sure that we never create a new socket when one socket already exists

Reply:
Ah thanks.